File: 2021/08/17/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, mx):
            if not node:
                return 0

            curr = (node.val >= mx)
            mx = max(mx, node.val)
            
            return dfs(node.left, mx) + dfs(node.right, mx) + curr

        return dfs(root, -1000001)

File: 2021/08/17/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func goodNodes(root *TreeNode) int {
	var dfs func(*TreeNode, int) int

	dfs = func(root *TreeNode, mx int) int {
		if root == nil {
			return 0
		}

		cur := 0
		if root.Val >= mx {
			cur = 1
			mx = root.Val
		}

		return dfs(root.Left, mx) + dfs(root.Right, mx) + cur
	}
	return dfs(root, 10000001)
}
